### ISSUE TITLE
Support `$/progress` notification from LSP 3.15.0

### ISSUE THREAD

**Issue by jez** [autozimu/LanguageClient-neovim#1031]
## Is your feature request related to a problem? Please describe.

The LSP spec, as of version 3.15.0, includes a way for a server to send arbitrary progress notifications from the server to the client:

<https://microsoft.github.io/language-server-protocol/specification.html#progress>

This generalizes ad-hoc features that have been added to LanguageClient-neovim to support similar functionality in RLS and cquery:

https://github.com/autozimu/LanguageClient-neovim/pull/243

The Sorbet (Ruby type checker) LSP server implementation has also had ad hoc support for these kinds of notifications, which it calls `sorbet/showOperation`:

https://github.com/sorbet/sorbet/blob/bb08e704f515ff7cd3f5e51da8aaca681cfa7f23/main/lsp/tools/make_lsp_types.cc#L1323

https://github.com/sorbet/sorbet/blob/bb08e704f515ff7cd3f5e51da8aaca681cfa7f23/main/lsp/tools/make_lsp_types.cc#L1252-L1258

## Describe the solution you'd like

I like LanguageClient-neovim to listen for `$/progress` notifications from the server, let a user to register arbitrary callback functions in Vim script to process these notifications. It would seem that we'd want to be able to register these per language, because the server gets to determine the contents of the message:

```ts
type ProgressToken = number | string;
interface ProgressParams<T> {
	/**
	 * The progress token provided by the client or server.
	 */
	token: ProgressToken;

	/**
	 * The progress data.
	 */
	value: T;
}
```

So something like

```vim
let g:LanguageClient_progressCallbacks = {}
let g:LanguageClient_progressCallbacks.ruby = s:MySorbetProgressCallback
```

This function would get the `ProgressParams` as a dictionary, or as close as we can get to it (I'm not sure whether all JSON-RPC-serialized values can be deserialized verbatim into a Vim dictionary; I'm fine dropping the keys / values that can't be, and providing a "best effort" because most servers are just going to be sending dicts / lists / strings / numbers there.

## Describe alternatives you've considered

The PR I linked above (#243) seems to write a value into a global variable. That seems less preferable to me than a progress callback, because with the callback approach, the user can decide to take action right away (like ring a bell or flash the screen, integrate with their status line, or even set a global variable).

## Additional context

I'm happy to work on supporting this if this is a feature that you're interested in accepting.

**Comment by martskins** [comment]
There's actually some ground covered on this in #1027 (which has been merged to dev already). But my assumptions on how to handle progress notifications may have been incorrect, so feel free to point out any problems you find with it.

I'm curious if you have seen any server side implementation of `$/progress` that would fall outside of the cases covered in that PR. I honestly couldn't find any example of any server that uses these notifications outside of `rust-analyzer`, which is what I used to test these, but it may very well be the case that other servers expect some other kind of treatment for these.
